src/contracts/ParaSwapDebtSwapAdapter.sol

```diff
@@ -141,9 +176,34 @@ abstract contract ParaSwapDebtSwapAdapter is
     require(msg.sender == address(POOL), 'CALLER_MUST_BE_POOL');
     require(initiator == address(this), 'INITIATOR_MUST_BE_THIS');
 
-    FlashParams memory swapParams = abi.decode(params, (FlashParams));
-    _swapAndRepay(swapParams, IERC20Detailed(assets[0]), amounts[0]);
+    FlashParams memory flashParams = abi.decode(params, (FlashParams));
+
+    // This is only non-zero if we flashed extra collateral.
+    if (flashParams.nestedFlashloanDebtAsset != address(0)) {
+      // Wrap the swap with a supply and withdraw.
+      address collateralAsset = assets[0];
+      uint256 collateralAmount = amounts[0];
+
+      // Supply
+      _supply(collateralAsset, collateralAmount, flashParams.user, REFERRER);
 
+      // Execute the nested flashloan
+      address newAsset = flashParams.nestedFlashloanDebtAsset;
+      flashParams.nestedFlashloanDebtAsset = address(0);
+      _flash(flashParams, newAsset, flashParams.nestedFlashloanDebtAmount);
+
+      // Fetch and transfer back in the aToken to allow the pool to pull it.
+      (, , address aToken) = _getReserveData(collateralAsset);
+      IERC20WithPermit(aToken).safeTransferFrom(flashParams.user, address(this), collateralAmount); // Could be rounding error but it's insignificant
+      POOL.withdraw(collateralAsset, collateralAmount, address(this));
+      _conditionalRenewAllowance(collateralAsset, collateralAmount);
+
+      // Return out of this scope.
+      return true;
+    }
+
+    // There is no need for additional collateral, execute the swap.
+    _swapAndRepay(flashParams, IERC20Detailed(assets[0]), amounts[0]);
     return true;
```

Comment:
I think will be more explicit to put else {} instead

```
      _conditionalRenewAllowance(collateralAsset, collateralAmount);
    } else {
      // There is no need for additional collateral, execute the swap.
      _swapAndRepay(flashParams, IERC20Detailed(assets[0]), amounts[0]);
    }
    return true;

Reply:
I like both approaches and am a fan of exiting scopes asap, but in this scenario completely indifferent, will change

Reply:
Addressed in https://github.com/bgd-labs/aave-debt-swap/pull/19/commits/31c93d01eead1ecd5d0d49de2090ad8af3aa37ed